diff --git a/third_party/mbedtls/repo.patched/library/ssl_srv.c b/third_party/mbedtls/repo.patched/library/ssl_srv.c
index 37f415dd1..cdbbaf3aa 100644
--- a/third_party/mbedtls/repo.patched/library/ssl_srv.c
+++ b/third_party/mbedtls/repo.patched/library/ssl_srv.c
@@ -3227,6 +3227,9 @@ curve_matching_done:
 
     /* Done with actual work; add header and send. */
 
+    // mask dead store of p
+    (void)p;
+
     ssl->out_msglen  = 4 + n;
     ssl->out_msgtype = MBEDTLS_SSL_MSG_HANDSHAKE;
     ssl->out_msg[0]  = MBEDTLS_SSL_HS_SERVER_KEY_EXCHANGE;
