diff --git a/third_party/mbedtls/repo.patched/library/ssl_srv.c b/third_party/mbedtls/repo.patched/library/ssl_srv.c
index fc0d2d7b..fad54dd8 100644
--- a/third_party/mbedtls/repo.patched/library/ssl_srv.c
+++ b/third_party/mbedtls/repo.patched/library/ssl_srv.c
@@ -3017,6 +3017,9 @@ curve_matching_done:
           MBEDTLS_KEY_EXCHANGE_ECDHE_RSA_ENABLED ||
           MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED */
 
+    // mask dead store of p
+    (void)p;
+
     ssl->out_msglen  = 4 + n;
     ssl->out_msgtype = MBEDTLS_SSL_MSG_HANDSHAKE;
     ssl->out_msg[0]  = MBEDTLS_SSL_HS_SERVER_KEY_EXCHANGE;
